packages/loader/container-loader/src/container.ts

```diff
@@ -1628,6 +1669,9 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
 
         const oldState = this._connectionState;
         this._connectionState = value;
+        // Set it to undefined in both "Disconnected" and "Connected" case as we only require it in
+        // transition from "Disconnected" to "Connected" and we cannot come here in "Connecting" event.
+        this.prevClientLeftP = undefined;
```

Comment:
This looks wrong to me. It does not account for situation where client may lose and gain connection couple times before it had a chance to see a leave op from very first connection. this.prevClientLeftP should be reset only on timeout or successful receival of join op. All other transitions should leave it as is!
This applies to all the logic (i.e. it's not about just this line of code).

Can we please add UTs to cover that and other cases? Consider (maybe as separate PR) to isolate all this logic into separate class, such that it's possible to test it in isolation without need to stand up full container

Reply:
Yes, just wanted to receive first feedback before writing tests/refactoring etc.